deps/sketch.py

```diff
@@ -0,0 +1,271 @@
+
+
+async def _install_libraries(  # pylint: disable=too-many-locals, too-many-branches
+    libraries: list[Library],
+) -> dict[Library, str]:
+    # Install required libraries
+    if not await check_for_internet():
+        logger.warning("No internet connection, skipping library install")
+        return {}
+
+    installed_library_versions = {}
+    for library in libraries:
+        library = library.strip()
+        potential_repos = library_indexed_json.get(library)
+        if not potential_repos:
+            raise HTTPException(404, f"Library {library} not found")
+
+        version_required = ""
+        if library.find("@") != -1:
+            version_required = library.split("@")[1]
+        else:
+            versions = [
+                re.sub(r"[^.0-9]", "", repo["version"]) for repo in potential_repos
+            ]
+            version_required = _get_latest_version(versions)
+
+        # Check if the library is already installed
+        if path.exists(f"./arduino-libs/{library}@{version_required}"):
+            installed_library_versions[library] = version_required
+            continue
+
+        logger.info("Installing libraries: %s@%s", library, version_required)
+
+        repo = list(
+            filter(
+                lambda x: x["version"]
+                == version_required,  # pylint: disable=cell-var-from-loop
+                potential_repos,
+            )
+        )[0]
+        if not repo:
+            raise HTTPException(
+                404, f"Library {library} not found, with version {version_required}"
+            )
+
+        library_dir = f"./arduino-libs/{repo['name']}@{repo['version']}"
+        async with httpx.AsyncClient() as client:
+            library_zip = zipfile.ZipFile(
+                io.BytesIO((await client.get(repo["url"])).content)
+            )
+        # All the content is in ZIP/ZIP_NAME/ but we want it in ZIP/ so we extract it to the root
+        # Only export any CPP files to lib/ dir
+        os.makedirs(library_dir, exist_ok=True)
+        os.mkdir(f"{library_dir}/src")
+
+        includes = {}
+        dir_deps = {}
+        arches, deps, deps_arches, in_deps = await _install_library_zip(
+            library_zip, repo, library_dir
+        )
+
+        for includes_board in fqbn_to_board.values():
+            includes[includes_board] = ""
+            dir_deps[includes_board] = ""
+
+        # Compile the library using platformio and store the compiled sources so we can use them later
+        if deps or (not "*" in arches):
+            for fqbn, board in fqbn_to_board.items():
+                if fqbn.split(":")[1] not in arches and "*" not in arches:
+                    continue
+                for dep in deps:
+                    if board not in deps_arches[dep] and "*" not in deps_arches[dep]:
+                        continue
+                    dir_path = f"../{dep}@{in_deps[dep]}/"
+                    dir_deps[board] += f"\t\t\t{dir_path}src\n"
```

Comment:
see above, why not just let platformio handle all this and use the regular deps mechanism. We're introducing platformio but don't actually use it? all this code seems fragile and messy to maintain. Esp. when considering this also has to run on a PI and in the desktop client. Less == more.